crates/sui-core/src/authority/authority_per_epoch_store.rs

```diff
@@ -2270,98 +2291,132 @@ impl AuthorityPerEpochStore {
+            .collect();
 
         // Load transactions deferred from previous commits.
-        // We do this after updating the last_randomness_round_written above so that every deferred
-        // transaction that can be run with this commit is loaded.
         let deferred_tx: Vec<VerifiedSequencedConsensusTransaction> = self
             .load_deferred_transactions_for_consensus_round(&mut batch, commit_round)?
             .into_iter()
-            .chain(self.load_deferred_transactions_for_randomness_round(
-                &mut batch,
-                last_randomness_round_written,
-            )?)
             .collect();
-        let previously_deferred_tx_digests: HashSet<_> = deferred_tx
-            .iter()
-            .map(|tx| match tx.0.transaction.key() {
-                SequencedConsensusTransactionKey::External(
-                    ConsensusTransactionKey::Certificate(digest),
-                ) => digest,
-                _ => panic!("deferred transaction was not a user certificate: {tx:?}"),
-            })
-            .collect();
-        sequenced_transactions.extend(deferred_tx.into_iter());
+        for tx in deferred_tx {
+            if tx.0.is_user_tx_with_randomness() {
+                sequenced_randomness_transactions.push(tx);
+            } else {
+                sequenced_transactions.push(tx);
+            }
+        }
 
+        // We always order transactions using randomness last.
         PostConsensusTxReorder::reorder(
             &mut sequenced_transactions,
             self.protocol_config.consensus_transaction_ordering(),
         );
+        PostConsensusTxReorder::reorder(
+            &mut sequenced_randomness_transactions,
+            self.protocol_config.consensus_transaction_ordering(),
+        );
         let consensus_transactions: Vec<_> = system_transactions
             .into_iter()
             .chain(sequenced_transactions)
+            .chain(sequenced_randomness_transactions)
             .collect();
 
-        let (transactions_to_schedule, notifications, lock, final_round) = self
+        let (transactions_to_schedule, notifications, lock, randomness_round, final_round) = self
             .process_consensus_transactions(
                 &mut batch,
                 &consensus_transactions,
                 &end_of_publish_transactions,
                 checkpoint_service,
                 cache_reader,
-                commit_round,
-                previously_deferred_tx_digests,
-                last_randomness_round_written,
             )
             .await?;
         self.record_consensus_commit_stats(&mut batch, consensus_stats)?;
 
-        // Create a pending checkpoint if we are still accepting tx.
+        // Create pending checkpoints if we are still accepting tx.
         let should_accept_tx = if let Some(lock) = &lock {
             lock.should_accept_tx()
         } else {
             self.get_reconfig_state_read_lock_guard().should_accept_tx()
```

Comment:
Please add comment on why it is safe to release the lock immediately.

Reply:
done